File: examples/shimdemoWLS.m
% demoWLS.m
%
% Demonstrate the proposed B0 shimming workflow using a synthesized baseline fieldmap f0.
% 
% Goal is to calculate shim setting changes s that minimize the weighted RMS B0 inhomogeneity.
% Provided here as an example; other loss functions may be more useful depending on the application.

% spherical harmonic basis degree
l = 6;

%% Calculate calibration matrix A
% This only needs to be done once per scanner

% load calibration data
load Ffull                             % [nx ny nz 8]  (does not include DC offset term)
[nx, ny, nz, nShim] = size(Ffull);
fov = 19.8*[1 1 1];   % cm

% generate mask (only for purpose of calculating A)
mask = sum(abs(Ffull),4) > 10;         % [nx ny nz]     
N = sum(mask(:));

% create F = [N 8] by applying mask 
F = zeros(N, nShim);
for ii = 1:nShim
	tmp = Ffull(:,:,:,ii);
	F(:,ii) = tmp(mask); 
end

% specify shim amplitudes (differences) used to obtain F (hardware units)
S = diag([20*ones(1,3) 100*ones(1,5)]);  % Do not include DC term. Hardware units (2nd order unit = 10 mA)

% Get spherical harmonic basis evaluated at the same N spatial locations as F
[X,Y,Z] = shim.getgrid(nx,ny,nz,fov);
H = shim.getSHbasis(X(mask),Y(mask),Z(mask),l);   % [N ...]

% Calculate calibration matrix A
% F and S do NOT include the DC term, i.e., F: [N 8], S: [8 8]
A = shim.getcalmatrix(F, H, S);


%% get shim amplitudes 'shat' that minimize RMS fieldmap
%shimNum = 5;
f0 = Ffull(:,:,:,shimNum);  % see how well one shim field fits basis
mask = sum(abs(Ffull),4) > 10;         % [nx ny nz]
f0 = f0(mask);   % [N 1]
W = diag_sp(ones(N,1));
shat = -(W*H*A)\(W*f0);    % [9 1]. NB! May need to be rounded before applying settings on scanner.

f = f0 + H*A*shat;  % predicted fieldmap after applying shims

f0 = embed(f0, mask);
f = embed(f, mask);

% compare baseline and predicted fieldmaps
subplot(121)
im(f0); 
%title(sprintf('fieldmaps: original (left in each panel),\nand after 2nd order shimming (right)'));
title(sprintf('before shimming'));
h = colorbar; h.TickLabels{end} = 'Hz'; % h.Label.String = 'B0 field (Hz)';
subplot(122)
im(f,[-5 5]); 
title(sprintf('Residual (shim term %d)', shimNum));
h = colorbar; h.TickLabels{end} = 'Hz';

File: julia/shimdemoWLS.jl
using JLD2
using MIRT: embed!, jim, ndgrid
using Random
using LinearAlgebra
using SparseArrays

include("getSHbasis.jl")
include("getcalmatrix.jl")
#include("ls_adam.jl")
include("shimoptim.jl")

# load calibration data
@load "CalibrationData.jld2" F S fov meta 

(nx,ny,nz,nShim) = size(F)

(x,y,z) = ndgrid(
	range(-fov[1], fov[1], length=nx), 
	range(-fov[2], fov[2], length=ny), 
	range(-fov[3], fov[3], length=nz) 
	)

# create mask
fm = sum(abs.(F), dims=4)[:,:,:,1]
mask = fm .> 200    # note the '.' (broadcasting)
mask[1:2:end, 1:2:end, 1:2:end] .= false
N = sum(mask[:])

# mask and reshape to [N 8] 
Fm = zeros(N, nShim)    # m for 'masked'
for ii = 1:nShim
	f1 = F[:,:,:,ii]
	Fm[:,ii] = f1[mask]
end

# Get spherical harmonic basis of degree l
l = 4
H = getSHbasis(x[mask], y[mask], z[mask], l)   # size is [N sum(2*(0:l) .+ 1)]

# Get calibration matrix A
A = getcalmatrix(Fm, H, S)

# Now that A is determined we can use it to optimize shims for a given acquired fieldmap.
# Example: Synthesize an example fieldmap 'f0' and optimize shims (minimize RMS residual) for that fieldmap.
f0 = sum(F[:,:,:,[2,6,7,8]], dims=4)[:,:,:,1]
f0 = F[:,:,:,2] + sqrt.(abs.(F[:,:,:,5])) + 3*F[:,:,:,6]
f0 = 2*F[:,:,:,2] + 10*sqrt.(abs.(F[:,:,:,5])) + 1*F[:,:,:,6]
f0 = F[:,:,:,2] + 2*F[:,:,:,6]
mask = abs.(f0) .> 0                # note the dots
mask[1:2:end, 1:2:end, 1:2:end] .= false
f0m = f0[mask]
N = sum(mask[:])
f0m = f0m + 0*randn(size(f0m))        # add some noise
H = getSHbasis(x[mask], y[mask], z[mask], l)  
W = sparse(collect(1:N), collect(1:N), ones(N))
W = Diagonal(ones(N,))

# Initial guess (unconstrained LS solution)
s0 = -(W*H*A)\(W*f0m)    # Vector of length 9. NB! May need to be rounded before applying settings on scanner.

# shim limits 
shimlims = (100, 2000, 12000)

# solve using Flux
# initialize with LS solution
(lin_max, hos_max, hos_sum_max) = shimlims
#shat[5:9] .= tan.(pi/2 * shat[5:9]/hos_max)   # initial guess
#@time (shat, out) = ls_adam(W*H*A, W*f0m; s0=s0, opt=opt, niter=niter)

# solve using NLopt
@time shat = shimoptim(W*H*A, W*f0m, shimlims) # ; s0=s0)

fpm = f0m + H*A*shat;      # predicted fieldmap after applying shims

# display predicted fieldmap
fp = zeros(size(f0))
embed!(fp, fpm, mask)   
p = jim(cat(f0,fp;dims=1))    # compare before and after shimming
p = jim(fp; clim=(-40,40))
#display(p)


